### PR TITLE
[Bugfix][Spec Decode] Only create draft_id_to_target_id when draft vocab differs

### PR DESCRIPTION
**Pull request by drslark**
## Purpose

We found a bug that makes full-vocab EAGLE3 drafters (`draft_vocab_size == target vocab_size`, e.g. `Inferact/MiniMax-M3-EAGLE3-GQA` and `nvidia/Qwen3-30B-A3B-Thinking-2507-Eagle3`) run a redundant vocab-remapping step after the lm_head, degrading decode performance with no effect on the output.

`compute_logits` in `llama_eagle3.py` has a fast path and a slow path. Full-vocab drafts were meant to take the fast path, but erroneously they never do. On H20, the slow path's `logits_new[:, targets] = logits` is `5x` slower than a plain copy:

```python
# vllm/model_executor/models/llama_eagle3.py::Eagle3LlamaForCausalLM
def compute_logits(self, hidden_states):
    logits = self.logits_processor(self.lm_head, hidden_states)
    if self.draft_id_to_target_id is None:
        # full-vocab draft: logits are already in target vocab space
        return logits

    # reduced-vocab draft: scatter into target vocab space
    base = torch.arange(self.config.draft_vocab_size, device=logits.device)
    targets = base + self.draft_id_to_target_id
    logits_new = logits.new_full(
        (logits.shape[0], self.config.vocab_size), float("-inf")
    )
    logits_new[:, targets] = logits
    return logits_new
```

In `__init__`, even when `draft_vocab_size == target vocab_size`, the parameter is still initialized — it is never `None`, so the slow path is always taken:

```python
# vllm/model_executor/models/llama_eagle3.py::Eagle3LlamaForCausalLM
def __init__(self, ...):
    ...
    self.draft_id_to_target_id = nn.Parameter(
        torch.zeros(self.config.draft_vocab_size, dtype=torch.long),
        requires_grad=False,
    )
    ...
```

After reading https://github.com/vllm-project/vllm/pull/18488 and https://github.com/vllm-project/vllm/pull/18909, we believe full-vocab drafts were always meant to take the fast path.

This is corroborated by `qwen3_dspark.py` and `qwen3_dflash.py`, which already handle it correctly:

```python
# vllm/model_executor/models/qwen3_dspark.py::Qwen3DSparkForCausalLM
def __init__(self, ...):
    ...
    target_vocab_size = vllm_config.model_config.get_vocab_size()
    if self.config.draft_vocab_size != target_vocab_size:
        self.draft_id_to_target_id = nn.Parameter(
            torch.zeros(self.config.draft_vocab_size, dtype=torch.long),
            requires_grad=False,
        )
    else:
        self.draft_id_to_target_id = None
    ...
```

So we follow dspark/dflash: create the parameter only when the draft vocab differs. Full-vocab drafts then take the `is None` fast path and return the lm_head logits as-is (bit-identical); reduced-vocab drafts are unaffected.

## Test Plan

1. Added a print to the slow path of `compute_logits` and ran a full-vocab EAGLE3 draft, before and after the change.
2. Added a unit test for the conditional: `pytest tests/model_executor/test_eagle_quantization.py -k d2t -v`
3. End-to-end speculative-decoding acceptance run on an H20 GPU with `nvidia/Qwen3-30B-A3B-Thinking-2507-Eagle3` (`num_speculative_tokens=3`), using 80 MT-Bench prompts passed through the chat template, before and after the change, measuring the acceptance rate.

## Test Result

1. Before the change, the print fires on every call; after the change, it never fires — the slow path is confirmed not taken, and the slow operation is avoided.
2. The unit test passes; the two full-vocab cases fail on unpatched code, so it guards the regression.
3. Before the change:

   ```
   total_num_output_tokens: 20480
   num_drafts: 12294
   num_draft_tokens: 36882
   num_accepted_tokens: 8171
   mean acceptance length: 1.66
   --------------------------------------------------
   acceptance at token 0: 0.41
   acceptance at token 1: 0.18
   acceptance at token 2: 0.08
   ```

   After the change:

   ```
   total_num_output_tokens: 20480
   num_drafts: 12294
   num_draft_tokens: 36882
   num_accepted_tokens: 8171
   mean acceptance length: 1.66
   --------------------------------------------------
   acceptance at token 0: 0.41
   acceptance at token 1: 0.18
   acceptance at token 2: 0.08
   ```

   Every counter matches exactly across both runs, confirming the change is output-neutral for full-vocab drafts.

This PR was prepared with AI assistance (Claude Code, GLM-5.3) and reviewed by the submitter.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by drslark** [comment]
@benchislett 
Since you authored https://github.com/vllm-project/vllm/pull/18488 — full-vocab EAGLE3 drafts still never take the None fast path. This PR makes draft_id_to_target_id conditional on vocab size (like dspark/dflash) to restore it. Could you take a look?

**Comment by drslark** [comment]
https://github.com/vllm-project/vllm/blob/0e14198a63c03f899a10f3e782e88eca7f11265b/vllm/v1/worker/gpu/spec_decode/speculator.py#L358-L388
@njhill 
I think the label is wrong, mrv2 will also call `logits = self.model.compute_logits(hidden_states)`.
This pr is also a bugfix for v2, too.

File: tests/model_executor/test_eagle_quantization.py [hunk 1]
```diff
@@ -158,3 +158,53 @@ def test_eagle3_lm_head_receives_quant_config():
...
+    ):
+        MockModel.return_value.use_aux_hidden_state = True
+
+        model = Eagle3LlamaForCausalLM(vllm_config=mock_vllm_config)
+
+    if expects_d2t:
+        assert isinstance(model.draft_id_to_target_id, torch.nn.Parameter)
+        assert model.draft_id_to_target_id.shape == (draft_vocab_size,)
+    else:
+        assert model.draft_id_to_target_id is None
```
Comment (Isotr0py) [inline tests/model_executor/test_eagle_quantization.py]:
```suggestion
```
Given that CI resources and maintainence effort are limited, no need to add tests for such a minor fix.

**Review by Isotr0py — APPROVED** [review]
